Address #54 substantial issues (#57)
greenelab/xswap-manuscript: content/02.body.md

Undirected edge types (e.g Compound–resembles–Compound) have only a single degree distribution.
Degree distributions are non-uniform and vary greatly between different networks.
The y-axis is log~10~-scaled to accommodate the common occurrence where most nodes have low degree while a small portion of nodes have high degree.
Several distributions have nodes that reach the maximum degree, corresponding to a node being connected to all other possible nodes.
Zero-degree nodes are not displayed, since methodological limitations often result in edge data only existing for a subset of nodes.
](https://github.com/greenelab/xswap-analysis/raw/b0db22c82b2e58bf1ef5ae78317167982016e26b/img/hetionet_degrees.png){#fig:hetionet width="100%"}

Degree is an important metric for differentiating between nodes, and it appears in many common edge prediction features [@doi:10.1155/2015/172879].
However, reliance on degree can pose problems for edge prediction.
First, bias in networks can distort node degree so that a difference in degree between two nodes in a given network may not reflect a true difference in number of relationships.

> Second, edge prediction methods relying heavily on degree may be nonspecific---predicting trivial rather than insightful new relationships.

Reviewer: ```suggestion
Second, edge prediction methods that rely heavily on degree may be nonspecific---predicting trivial rather than insightful new relationships.
```

Another reason why a reliance on degree can be unfavorable is that degree imbalance can lead to prediction nonspecificity.
Nonspecific predictions are not made on the basis of the specific connectivity information contained in a network.
For example, Gillis et al. [@doi:10.1371/journal.pone.0017258] examined the concept of prediction specificity in the context of gene function prediction and found that many predictions appear to rely primarily on multifunctionality and could be "potentially misleading with respect to causality."
Degree imbalance leads high-degree nodes to dominate in the predictions made by degree-associated methods [@doi:10.1093/bioinformatics/btv215], which are effective predictors of connections in some biological networks [@doi:10.1186/1752-0509-2-11].
Consequently, degree-based predictions are more likely nonspecific, meaning the same set of predictions performs well for different tasks.

Depending on the prediction task, edge predictions involving very high degree nodes may be undesired, uninsightful, or nonspecific.

> While nonspecificity may be acceptable in some cases, we believe that specific predictions are desirable for a range of prediction tasks in biomedicine as they represent insight gained from the actual connections in a network.

Reviewer: I'm not sure I love this wording. I think what you're going for is that you're more interested in predictions that come from the nodes that are connected. Even what you're calling non-specific here is driven by the actual connections in the network (degree is a function of this).

Author: ```suggestion
While nonspecific predictions may be acceptable in some cases, we believe that specific predictions are desirable for a range of prediction tasks in biomedicine as they represent insight gained from the particular connections in a network.
```
I'm trying to explain what I mean by specificity without using the word 'specific', and acknowledge that nonspecificity isn't always a problem. Would this be more clear?

Reviewer: I think for the latter part you want it to be about the nodes that are connected by edges. I _think_ that's what you mean.

Author: That is what I mean. Predictions should reflect the edges between nodes (though obviously we are not predicting connections between directly connected egdes).
```suggestion
While nonspecificity may be acceptable in some cases, we believe that predictions based on the edges and network structure connecting two nodes are desirable for a range of prediction tasks in biomedicine.
```

Reviewer 2: I think all the potential phrasings are okay. I see why it's a hard concept to communicate succinctly. Feel free to use any of the following for inspiration.

```suggestion
While predictions based on primarily on degree may be acceptable for some tasks,
generating less obvious insights from networks requires drawing inferences from particular connections, where analysis considers both the source and target node of an edge.
```

Author: I like the suggestion! How about merging the two phrasings like this?
```suggestion
While predictions based primarily on degree may be acceptable for some tasks, generating less obvious insights from networks requires drawing inferences from the specific connections and network structure between nodes.
```

Reviewer 2: > How about merging the two phrasings like this?

looks good to me

Undirected edge types (e.g Compound–resembles–Compound) have only a single degree distribution.
Degree distributions are non-uniform and vary greatly between different networks.
The y-axis is log~10~-scaled to accomodate the common occurrence where most nodes have low degree while a small portion of nodes have high degree.

> Several distributions have node degree ceilings corresponding to a node being connected to all other possible nodes.

Author: Commentary:
> odd phrase

For a given node pair, degree grouping treats other node pairs with the same degrees as additional permutations.
We used this strategy to augment the number of predictor values for each node pair in permuted networks, allowing node pairs to have more permuted predictor values than permuted networks.
Degree grouping greatly increased the effective number of permutations for nodes with frequently observed degrees [@url:https://github.com/greenelab/hetmech/pull/96].
We used degree grouping throughout our analyses.

### Implementation and source code

We implemented our modified version of the XSwap algorithm as an open-source Python package.
The package contains modules for permuting networks, computing the edge prior, and converting networks between adjacency matrix and edge list formats.
Additionally, we include the analytical approximation of the edge prior and functionality to assign unique identifiers to nodes.
The Python package is [available](https://pypi.org/project/xswap/) on the Python Packaging Index under the name "xswap".

> The full source code has also been made freely available under the BSD 2-Clause License (<https://github.com/hetio/xswap>).

Reviewer: ```suggestion
The full source code is freely available under the BSD 2-Clause License (<https://github.com/hetio/xswap>).
```
